test/e2e/storage/testsuites/snapshottable_metrics.go

```diff
@@ -0,0 +1,400 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testsuites
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	"github.com/prometheus/common/model"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2emetrics "k8s.io/kubernetes/test/e2e/framework/metrics"
+	e2epv "k8s.io/kubernetes/test/e2e/framework/pv"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	e2evolume "k8s.io/kubernetes/test/e2e/framework/volume"
+	storageframework "k8s.io/kubernetes/test/e2e/storage/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	storageutils "k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+type snapshottableMetricsTestSuite struct {
+	tsInfo storageframework.TestSuiteInfo
+}
+
+var _ storageframework.TestSuite = &snapshottableMetricsTestSuite{}
+
+// InitSnapshottableMetricsTestSuite returns snapshottableMetricsTestSuite that implements TestSuite interface
+func InitSnapshottableMetricsTestSuite() storageframework.TestSuite {
+	return &snapshottableMetricsTestSuite{
+		tsInfo: storageframework.TestSuiteInfo{
+			Name: "snapshottableMetrics",
+			TestPatterns: []storageframework.TestPattern{
+				storageframework.DynamicSnapshotDelete,
+				storageframework.PreprovisionedSnapshotDelete,
+			},
+			SupportedSizeRange: e2evolume.SizeRange{
+				Min: "1Mi",
+			},
+			FeatureTag: "[Feature:VolumeSnapshotDataSource]",
+		},
+	}
+}
+
+func (s *snapshottableMetricsTestSuite) GetTestSuiteInfo() storageframework.TestSuiteInfo {
+	return s.tsInfo
+}
+
+func (s *snapshottableMetricsTestSuite) SkipUnsupportedTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
```

Comment:
If a cluster is running a snapshot controller that is not visible as a Pod, how do we skip these tests?

Reply:
Similarly, this test is not going to work if someone runs snapshot controller with a different metrics port?

Reply:
Since these tests are more so for the snapshot-controller, I'm thinking we skip this suite unless we're testing the hostpath CSI driver? Other storage e2e tests are testing

Reply:
Whether or not to run this test depends more on the cluster than the driver. You can run hostpath tests on a cluster where snapshot-controller is not visible. Since this test not really testing the csi driver, I wonder if we should make this a more targeted test in the external-snapshotter repo instead of the testsuites here? We'll have control over the cluster that the test runs on.

Reply:
Makes sense. @xing-yang, should we move this test suite to the external-snapshotter repo? 

I suppose we'd need a new e2e-type suite that starts a kind cluster and validates snapshot-controller logic like the metrics feature.

Reply:
I wonder if it is possible to modify that e2e test for it to be more flexible so it would work for other clusters?  Another thing is that if we are moving it out, should we actually have a separate repo for all out-of-tree e2e tests?  This needs a e2e framework out of tree so it is not just for the external-snapshotter repo.

Reply:
Discussed offline, we will move this test to [csi_mock](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/csi_mock_volume.go) and make it configurable.